core/integration/kafka/plugin.go

```diff
@@ -0,0 +1,271 @@
+	bytes := []byte(detectorsParam)
+	err = json.Unmarshal(bytes, &detectorsSlice)
+	if err != nil {
+		log.WithError(err).
+			Error("error processing the detectors list")
+		return
+	}
+	return detectorsSlice, nil
+}
+
+func (p *Plugin) NewEnvStateObject(varStack map[string]string, state string) *kafkapb.EnvInfo {
+	envId, ok := varStack["environment_id"]
+	if !ok {
+		log.Error("cannot acquire environment ID")
+		return nil
+	}
+
+	var runNumberOpt *uint32 = nil
+	var runTypeOpt *string = nil
+	if state == "RUNNING" {
+		// the following fields are relevant only in RUNNING state
+		runNumberStr, ok := varStack["run_number"]
+		if ok {
+			runNumber, err := strconv.ParseUint(runNumberStr, 10, 32)
+			if err == nil {
+				runNumber32 := uint32(runNumber)
+				runNumberOpt = &runNumber32
+			}
+		}
+
+		runType, ok := varStack["run_type"]
+		if ok {
+			runTypeOpt = &runType
+		}
+	}
+
+	detectorsStr, ok := varStack["detectors"]
+	if !ok {
+		log.WithField("partition", envId).
+			Error("cannot acquire general detector list from varStack")
+	}
+	detectorsSlice, err := parseDetectors(detectorsStr)
+	if err != nil {
+		log.WithField("partition", envId).
+			Error("cannot parse general detector list")
+		return nil
+	}
+
+	return &kafkapb.EnvInfo{
+		EnvironmentId: envId,
+		RunNumber:     runNumberOpt,
+		RunType:       runTypeOpt,
+		State:         state,
+		Detectors:     detectorsSlice,
+	}
+}
+
+func (p *Plugin) UpdateRunningEnvList(envInfo *kafkapb.EnvInfo) {
+	if envInfo.State == "RUNNING" {
+		p.envsInRunning[envInfo.EnvironmentId] = envInfo
+	} else {
+		delete(p.envsInRunning, envInfo.EnvironmentId)
+	}
+}
+
+func (p *Plugin) GetRunningEnvList() []*kafkapb.EnvInfo {
+	var array []*kafkapb.EnvInfo
+	for _, v := range p.envsInRunning {
+		array = append(array, v)
+	}
+	return array
+}
+
+func (p *Plugin) ProduceMessage(message []byte, topic string, envId string) {
+	log.WithField("partition", envId).
+		Debug("Producing a new kafka message on topic ", topic)
+
+	deliveryChannel := make(chan kafka.Event)
+	err := p.kafkaProducer.Produce(&kafka.Message{
+		TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
+		Value:          message,
+	}, deliveryChannel)
+
+	if err != nil {
+		log.WithField("partition", envId).
+			Error("Kafka message delivery failed: ", err.Error())
+	}
+
+	e := <-deliveryChannel
+	m := e.(*kafka.Message)
+
+	if m.TopicPartition.Error != nil {
+		log.WithField("partition", envId).Error("Kafka message delivery failed: ", m.TopicPartition.Error)
+	} else {
+		log.WithField("partition", envId).Debugf("Kafka message delivered message to topic %s [%d] at offset %v\n",
+			*m.TopicPartition.Topic, m.TopicPartition.Partition, m.TopicPartition.Offset)
+	}
+}
+
+func (p *Plugin) CreateUpdateCallback(varStack map[string]string, state string) func() string {
+	return func() (out string) {
+		// Retrieve and update the env info
+		timestamp := uint64(time.Now().UnixNano() / 1000000)
```

Comment:
We now depend on Go 1.17 or later, so it'd be best to use `.UnixMilli()` for readability